elasticapm/contrib/sanic/__init__.py

```diff
@@ -0,0 +1,338 @@
+        :param handled: Boolean indicator for if the exception is handled.
+        :param kwargs: additional context to be passed to the API client for capturing exception related information
+        :return: None
+        """
+        assert self._client, "capture_exception called before application configuration is initialized"
+        return self._client.capture_exception(exc_info=exc_info, handled=handled, **kwargs)
+
+    async def capture_message(self, message=None, param_message=None, **kwargs):
+        """
+        Capture a generic message for the APM Client
+        :param message: Message information to be captured
+        :param param_message:
+        :param kwargs: additional context to be passed to the API client for capturing exception related information
+        :return:
+        """
+        assert self._client, "capture_message called before application configuration is initialized"
+        return self._client.capture_message(message=message, param_message=param_message, **kwargs)
+
+    def _setup_client_config(self, config: APMConfigType = None):
+        app_based_config = SanicAPMConfig(self._app)
+        if dict(app_based_config):
+            self._client_config = dict(app_based_config)
+
+        if config:
+            self._client_config.update(config)
+
+    # noinspection PyBroadException,PyUnresolvedReferences
+    def _init_app(self) -> None:
+        """
+        Initialize all the required middleware and other application infrastructure that will perform the necessary
+        capture of the APM instrumentation artifacts
+        :return: None
+        """
+        if not self._client:
+            self._client = make_client(config=self._client_config, client_cls=self._client_cls, **self._client_config)
+
+        if not self._skip_init_exception_handler:
+            self._setup_exception_manager()
+
+        if self._client.config.instrument and self._client.config.enabled:
+            instrument()
+            try:
+                from elasticapm.contrib.celery import register_instrumentation
+
+                register_instrumentation(client=self._client)
+            except ImportError:
+                self._logger.debug(
+                    "Failed to setup instrumentation. "
+                    "Please install requirements for elasticapm.contrib.celery if instrumentation is required"
+                )
+                pass
+
+        if not self._skip_init_middleware:
+            self._setup_request_handler(entity=self._app)
+
+    # noinspection PyMethodMayBeStatic
+    def _default_transaction_name_generator(self, request: Request) -> str:
+        """
+        Method used to extract the default transaction name. This is generated by joining the HTTP method and the
+        URL path used for invoking the API handler
+        :param request: Sanic HTTP Request object
+        :return: string containing the Transaction name
+        """
+        return f"{request.method} {request.path}"
```

Comment:
Let me quickly test the difference in overhead caused by this change of moving from current to anything else that can get the url as `/books/<id>`. However, if you notice, there is a way to customize this behavior that is provided to the end user already using `transaction_name_callback` argument while creating the `ElasticAPM` instance.

Reply:
Let me use `request.uri_template` instead. This give me `'/stuff/<info:int>'` format for the URL. Do you think I should strip the `:int` or that is good to have ?

Reply:
I am using request.uri_template in my custom instrumentation and it looks good.  I would personally like to have param type

Reply:
Reopening this item back again. The `uri_template` has a small behavior that makes it not really suitable. I did a quick test to confirm my suspicion. It breaks the `span` generation workflow. Will find a better suitable alternative.

Reply:
I will move to using the one suggested by @beniwohli  but in a sort of different way. 

```python

Reply:
```python
        url_template = request.path
        # Sanic's new router puts this into the request itself so that it can be accessed easily
        if hasattr(request, "route"):
            url_template = request.route.path
        else:
            # Let us fallback to using old router model to extract the info
            try:
                _, _, _, url_template, _, _ = self._app.router.get(request=request)
            except:
                pass

        return f"{request.method} {url_template}"
```
Falling back to this way to make sure we are compatible with both version of `sanic` as the revamped router has changed a bit in how the data is accessed.

@ahopkins Does this look all right ?

Reply:
One quick thing worth mentioning, the `request.route.path` string will not have the first `/` in the path. Not sure if this matters for your usage.